Firestore/core/src/firebase/firestore/core/filter.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_FILTER_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_FILTER_H_
+
+#include <memory>
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/model/document.h"
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+/** Interface used for all query filters. All filters are immutable. */
+class Filter {
+ public:
+  enum class Operator {
+    LessThan,
+    LessThanOrEqual,
+    Equal,
+    GreaterThan,
+    GreaterThanOrEqual,
+  };
+
+  /**
+   * Creates a Filter instance for the provided path, operator, and value.
+   *
+   * Note that if the relational operator is Equal and the value is NullValue or
+   * NaN, then this will return the appropriate NullFilter or NaNFilter class
+   * instead of a RelationFilter.
+   */
+  static std::unique_ptr<Filter> Create(model::FieldPath path,
```

Comment:
I see essentially no value in making this more general than it needs to be. We expect to use `std::shared_ptr<Filter>` everywhere so there's no reason to create these as `std::unique_ptr` and then convert literally everywhere we use them.

There's substantial benefit to making these `std::shared_ptr`s from the get-go: `std::make_shared<RelationFilter>()` only does a single allocation for both the shared_ptr control block and the object itself.

Meanwhile if you allocate them as unique_ptrs and then convert you end up having to allocate the control block separately.

However, see other comments.

Reply:
Done.